src/pykeen/datasets/utils.py

```diff
@@ -0,0 +1,225 @@
+# -*- coding: utf-8 -*-
+
+"""Dataset utilities."""
+
+import base64
+import hashlib
+import logging
+import pathlib
+from typing import Any, Collection, Iterable, Mapping, Optional, Pattern, Tuple, Type, Union
+
+import click
+import docdata
+from tqdm import tqdm
+
+from .base import Dataset, EagerDataset, PathDataset
+from ..constants import PYKEEN_DATASETS
+from ..triples import CoreTriplesFactory
+
+logger = logging.getLogger(__name__)
+
+dataset_regex_option = click.option("--dataset-regex", help="Regex for filtering datasets by name")
+max_triples_option = click.option("--max-triples", type=int)
+min_triples_option = click.option("--min-triples", type=int)
+
+
+def iter_dataset_classes(
+    regex_name_filter: Union[None, str, Pattern] = None,
+    *,
+    max_triples: Optional[int] = None,
+    min_triples: Optional[int] = None,
+    use_tqdm: bool = True,
+) -> Iterable[Tuple[str, Type[Dataset]]]:
+    """Iterate over dataset classes with given constraints.
+
+    :param regex_name_filter: An optional regular expression string or pre-compiled regular expression
+    :param max_triples: An optional maximum number of triples for the dataset
+    :param min_triples: An optional minimum number of triples for the dataset
+    :param use_tqdm: Should a progress bar be shown?
+    :yields: Pairs of dataset names and dataset classes
+    """
+    from . import dataset_resolver
+
+    it = sorted(
+        dataset_resolver.lookup_dict.items(),
+        key=Dataset.triples_pair_sort_key,
+    )
+    if max_triples is not None:
+        it = [pair for pair in it if Dataset.triples_pair_sort_key(pair) <= max_triples]
+    if min_triples is not None:
+        it = [pair for pair in it if Dataset.triples_pair_sort_key(pair) >= min_triples]
+    if regex_name_filter is not None:
+        if isinstance(regex_name_filter, str):
+            import re
+
+            regex_name_filter = re.compile(regex_name_filter)
+        it = [(name, dataset) for name, dataset in it if regex_name_filter.match(name)]
+    it_tqdm = tqdm(
+        it,
+        desc="Datasets",
+        disable=not use_tqdm,
+    )
+    for k, v in it_tqdm:
```

Comment:
```suggestion
    for name, dataset_cls in it_tqdm:
```
(while reviewing I had to search for what is contained in `it`)

Reply:
2719de4b